src/rulesets/reader.ts

```diff
@@ -45,28 +50,64 @@ async function readRulesFromRuleset(
     },
   });
   const ruleset = assertValidRuleset(JSON.parse(JSON.stringify(result)));
-
-  const newRuleset: IRulesetFile = {
-    rules: {},
+  const rules = {};
+  const functions = {};
+  const newRuleset: IRuleset = {
+    rules,
+    functions,
   };
 
   const extendedRulesets = ruleset.extends;
+  const rulesetFunctions = ruleset.functions;
 
-  if (extendedRulesets !== undefined) {
+  if (extendedRulesets !== void 0) {
     for (const extended of Array.isArray(extendedRulesets) ? extendedRulesets : [extendedRulesets]) {
+      let extendedRuleset: IRuleset;
+      let parentSeverity: FileRulesetSeverity;
       if (Array.isArray(extended)) {
-        const parentSeverity = severity === undefined ? extended[1] : severity;
-        mergeRulesets(newRuleset, await readRulesFromRuleset(uriCache, uri, extended[0], parentSeverity), {
-          severity: parentSeverity,
-        });
+        parentSeverity = severity === undefined ? extended[1] : severity;
+        extendedRuleset = await processRuleset(uriCache, uri, extended[0], parentSeverity);
       } else {
-        const parentSeverity = severity === undefined ? 'recommended' : severity;
-        mergeRulesets(newRuleset, await readRulesFromRuleset(uriCache, uri, extended, parentSeverity), {
-          severity: parentSeverity,
-        });
+        parentSeverity = severity === undefined ? 'recommended' : severity;
+        extendedRuleset = await processRuleset(uriCache, uri, extended, parentSeverity);
       }
+
+      mergeRules(rules, extendedRuleset.rules, parentSeverity);
+      mergeFunctions(functions, extendedRuleset.functions, rules);
     }
   }
 
-  return mergeRulesets(newRuleset, ruleset);
+  mergeRules(rules, ruleset.rules);
+  if (Array.isArray(ruleset.formats)) {
+    mergeFormats(rules, ruleset.formats);
+  }
+
+  if (rulesetFunctions !== void 0) {
+    const rulesetFunctionsBaseDir = join(
+      rulesetUri,
+      ruleset.functionsDir !== void 0 ? join('..', ruleset.functionsDir) : '../functions',
+    );
+    const resolvedFunctions: FunctionCollection = {};
+
+    await Promise.all(
+      rulesetFunctions.map(async fn => {
```

Comment:
does this map with promises work? I seem to remember failing with this and seeing articles about it not working as expected, but im certainly not an expert on this.

Reply:
Yeah, in general you are right.
If any promise fails, the other might not be executed at all, and the whole `Promise.all` promise is rejected.

I am not really sure how we want to handle that case.
If one function cannot be loaded, some rules won't work correctly, yet the others should work just fine, so there is no need to throw entirely.
That said, we might want to wrap each async function here `rulesetFunctions.map(async fn => {`  with a try catch.

Reply:
https://github.com/stoplightio/spectral/pull/430/commits/7a99652ae4de81532212f7413597648421a3791b